File: android/app/src/main/java/com/rork/guidestreamtvandroid/ui/components/PushReauthBanner.kt
package com.rork.guidestreamtvandroid.ui.components

import android.Manifest
import android.content.Context
import android.os.Build
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.rork.guidestreamtvandroid.data.repository.AuthViewModel
import com.rork.guidestreamtvandroid.data.repository.PushTokenManager
import com.rork.guidestreamtvandroid.ui.theme.BrandOrange
import com.rork.guidestreamtvandroid.ui.theme.TextPrimary
import com.rork.guidestreamtvandroid.ui.theme.TextSecondary

private const val PREFS_NAME = "gs_prefs"
private const val DISMISSED_KEY = "gs.pushReauthBannerDismissed"

private fun bannerDismissed(context: Context): Boolean =
    context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        .getBoolean(DISMISSED_KEY, false)

private fun setBannerDismissed(context: Context) {
    context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        .edit()
        .putBoolean(DISMISSED_KEY, true)
        .apply()
}

/**
 * Home-screen prompt shown when the account wants push
 * (`users.notify_push`, restored from the server) but POST_NOTIFICATIONS has
 * never been asked for on this install.
 *
 * GUI-41 put this message on Profile → Notifications only. That screen is
 * reached deliberately, and someone who just reinstalled — or who signed in on
 * a new phone — has no reason to go looking; they would simply stop getting
 * alerts and never find out why. Home is where they already are.
 *
 * Renders nothing when the permission is [NotificationPermissionState.GRANTED]
 * (including every device below Android 13, where no runtime grant exists) or
 * [NotificationPermissionState.DENIED] — a denial is the user's own decision,
 * Android will not re-show the dialog for it, and the settings screen already
 * carries the "open Settings" banner for that case.
 *
 * Dismissal is stored in SharedPreferences, which Android wipes along with the
 * app, so "not now" holds for this install and a later reinstall asks again.
 */
@Composable
fun PushReauthBanner(modifier: Modifier = Modifier) {
    val context = LocalContext.current
    val authVm = AuthViewModel.get()
    val notifyPush by authVm.notifyPushEnabled.collectAsStateWithLifecycle()

    var permState by remember { mutableStateOf(notificationPermissionState(context)) }
    var dismissed by remember { mutableStateOf(bannerDismissed(context)) }

    val lifecycleOwner = LocalLifecycleOwner.current
    DisposableEffect(lifecycleOwner) {
        // Catches a grant made in system Settings while the app was
        // backgrounded, so the banner disappears on return.
        val observer = LifecycleEventObserver { _, event ->
            if (event == Lifecycle.Event.ON_RESUME) {
                permState = notificationPermissionState(context)
            }
        }
        lifecycleOwner.lifecycle.addObserver(observer)
        onDispose { lifecycleOwner.lifecycle.removeObserver(observer) }
    }

    val permissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission(),
    ) { granted ->
        markNotificationPermissionAsked(context)
        permState = notificationPermissionState(context)
        authVm.setNotifyPushEnabled(granted)
        if (granted) {
            PushTokenManager.get().registerIfPermitted()
        } else {
            // They were asked and said no. Stop showing it; the settings
            // screen still explains how to turn notifications back on.
            setBannerDismissed(context)
            dismissed = true
        }
    }

    val visible = notifyPush &&
        permState == NotificationPermissionState.NOT_DETERMINED &&
        !dismissed
    if (!visible) return

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(14.dp))
            .background(BrandOrange.copy(alpha = 0.10f))
            .border(1.dp, BrandOrange.copy(alpha = 0.30f), RoundedCornerShape(14.dp))
            .padding(14.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Icon(
            imageVector = Icons.Filled.Notifications,
            contentDescription = null,
            tint = BrandOrange,
            modifier = Modifier.size(18.dp).padding(top = 1.dp),
        )

        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = "Turn notifications back on",
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold,
                color = TextPrimary,
            )
            Spacer(Modifier.height(3.dp))
            Text(
                text = "Push is off on this device. Your alert types are still saved.",
                fontSize = 12.sp,
                color = TextSecondary,
            )
            Spacer(Modifier.height(10.dp))
            Box(
                modifier = Modifier
                    .clip(CircleShape)
                    .background(BrandOrange.copy(alpha = 0.85f))
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                    ) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                            permissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                        }
                    }
                    .padding(horizontal = 14.dp, vertical = 7.dp),
            ) {
                Text(
                    text = "Turn back on",
                    fontSize = 13.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = Color.White,
                )
            }
        }

        Box(
            modifier = Modifier
                .size(28.dp)
                .clip(CircleShape)
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                ) {
                    setBannerDismissed(context)
                    dismissed = true
                },
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                imageVector = Icons.Filled.Close,
                contentDescription = "Dismiss",
                tint = TextSecondary,
                modifier = Modifier.size(14.dp),
            )
        }
    }
}

File: ios/GuideStreamTV/Views/PushReauthBanner.swift
//
//  PushReauthBanner.swift
//  GuideStreamTV
//
//  Home-screen prompt for the one state where notifications are off through
//  no decision of the user's: the account still wants push
//  (`users.notify_push = true`, restored from the server) but iOS reports
//  `.notDetermined`. Deleting an app revokes its notification authorization,
//  and a device the account has never run on never had one — so the copy says
//  "off on this device" rather than blaming a delete, which would be wrong in
//  the second case.
//
//  GUI-41 originally put this message on Profile → Notifications. That screen
//  is only ever reached deliberately, and someone who just reinstalled has no
//  reason to go looking — they would simply stop receiving alerts and never
//  find out why. Home is where they already are, so the prompt belongs here.
//  The settings banner stays as the explanation for anyone who does go looking.
//
//  Deliberately narrow. It does not appear on a new install (intent is false
//  until the server says otherwise, and onboarding does the first ask), and it
//  does not appear when the user turned notifications off in iOS Settings
//  themselves — that is `.denied`, a decision worth respecting, and iOS will
//  not re-prompt for it anyway.
//
//  Split into state + view so Home can write `if prompt.isVisible { … }`. A
//  self-contained view that returns an empty body still counts as a child of
//  Home's `VStack(spacing: 20)` and would leave a 20pt gap above the search
//  bar on every launch where the banner is not showing.
//

import SwiftUI
import UserNotifications
import UIKit

// MARK: - State

@Observable
@MainActor
final class PushReauthPrompt {
    static let shared = PushReauthPrompt()

    /// Dismissal is stored in UserDefaults, which iOS wipes along with the
    /// app — so "not now" holds for this install, and a later reinstall asks
    /// again, which is exactly the case this prompt exists for.
    private static let dismissedKey = "gs.pushReauthBannerDismissed"

    private(set) var isVisible: Bool = false
    private(set) var isRequesting: Bool = false

    private var dismissed: Bool {
        get { UserDefaults.standard.bool(forKey: Self.dismissedKey) }
        set { UserDefaults.standard.set(newValue, forKey: Self.dismissedKey) }
    }

    private init() {}

    /// Reads the same two facts as `NotificationsSettingsView`'s
    /// `refreshSystemStatus()` — saved intent and the live iOS grant — so the
    /// two surfaces can never disagree about whether push is really on.
    func refresh() async {
        let status = await UNUserNotificationCenter.current()
            .notificationSettings().authorizationStatus
        let wants = AuthViewModel.shared.notifyPushEnabled
        isVisible = wants && status == .notDetermined && !dismissed
    }

    func dismiss() {
        dismissed = true
        isVisible = false
    }

    /// Same path as the settings master toggle: ask iOS, and on a grant
    /// register for remote notifications so `push_tokens` picks up this
    /// install's APNs token instead of the deleted one's dead token.
    func requestAuthorization() async {
        guard !isRequesting else { return }
        isRequesting = true
        defer { isRequesting = false }

        let granted = (try? await UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .badge, .sound])) ?? false
        if granted {
            UIApplication.shared.registerForRemoteNotifications()
            AuthViewModel.shared.setNotificationPreferences(push: true, sms: false)
            isVisible = false
        } else {
            // They were asked and said no. Record that as their intent and stop
            // asking; Profile → Notifications still explains how to turn it on.
            AuthViewModel.shared.setNotificationPreferences(push: false, sms: false)
            dismiss()
        }
    }
}

// MARK: - Observation

private struct PushReauthObserver: ViewModifier {
    @Environment(\.scenePhase) private var scenePhase
    @State private var auth = AuthViewModel.shared

    func body(content: Content) -> some View {
        content
            .task { await PushReauthPrompt.shared.refresh() }
            // Permission can be granted in iOS Settings without the button
            // ever being tapped, so re-check on every return to the front.
            .onChange(of: scenePhase) { _, phase in
                guard phase == .active else { return }
                Task { await PushReauthPrompt.shared.refresh() }
            }
            // Intent arrives from the server after sign-in, which usually
            // lands after Home has already appeared.
            .onChange(of: auth.notifyPushEnabled) { _, _ in
                Task { await PushReauthPrompt.shared.refresh() }
            }
    }
}

extension View {
    /// Keeps `PushReauthPrompt.shared` current. Apply once, to a view that
    /// stays mounted for as long as Home does — not to the banner itself,
    /// which is conditional.
    func pushReauthObserver() -> some View {
        modifier(PushReauthObserver())
    }
}

// MARK: - View

struct PushReauthBanner: View {
    @State private var prompt = PushReauthPrompt.shared

    var body: some View {
        HStack(alignment: .top, spacing: 10) {
            Image(systemName: "bell.badge.fill")
                .font(.system(size: 16, weight: .semibold))
                .foregroundStyle(Color.orange)
                .padding(.top, 1)

            VStack(alignment: .leading, spacing: 3) {
                Text("Turn notifications back on")
                    .scaledFont(size: 14, weight: .bold)
                    .foregroundStyle(.white)
                Text("Push is off on this device. Your alert types are still saved.")
                    .scaledFont(size: 12)
                    .foregroundStyle(Color.textSecondary)
                    .fixedSize(horizontal: false, vertical: true)

                Button {
                    Task { await prompt.requestAuthorization() }
                } label: {
                    Text(prompt.isRequesting ? "Turning on…" : "Turn back on")
                        .scaledFont(size: 13, weight: .semibold)
                        .foregroundStyle(.white)
                        .padding(.horizontal, 14)
                        .padding(.vertical, 7)
                        .background(Capsule().fill(Color.orange.opacity(0.85)))
                }
                .buttonStyle(.plain)
                .disabled(prompt.isRequesting)
                .padding(.top, 6)
            }

            Spacer(minLength: 0)

            Button {
                UIImpactFeedbackGenerator(style: .light).impactOccurred()
                prompt.dismiss()
            } label: {
                Image(systemName: "xmark")
                    .font(.system(size: 11, weight: .bold))
                    .foregroundStyle(Color.textSecondary)
                    .padding(6)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Dismiss")
        }
        .padding(14)
        .background(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .fill(Color.orange.opacity(0.10))
        )
        .overlay(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .stroke(Color.orange.opacity(0.30), lineWidth: 1)
        )
    }
}
